firebase/firebase-js-sdk
# Typings Refactor

This PR refactors the single top level `firebase` package curated types into a series of 6 "types packages."

They are:

- `@firebase/app-types`
- `@firebase/auth-types`
- `@firebase/database-types`
- `@firebase/firestore-types`
- `@firebase/messaging-types`
- `@firebase/storage-types`

Each of these "types packages" defines the curated public interface and is used to generate our typings for the rest of the SDK.

Each component is encouraged to `implement`/`extend` these interfaces in their implementation directly (see the `firestore` package for a great example of this being done consistently throughout a codebase). Allowing the type system to validate that we are in fact exposing the interface we think we are.

## Review Guide

All reviewers should take a look at the following commits as it pertains to their component (i.e. look at `packages/app`, `packages/<COMPONENT>`, `packages/<COMPONENT>-types`, and `packages/firebase/<COMPONENT>` in each of these commits:

- d85410f - @firebase/app type refactor (`@firebase/auth` reviewer can skip this section)
- 9786607 - Adding the typings fields
- 51cff8d - Adding type package deps
- 086971e - Changes to the top level package

Broken down for each package owner are the package specific changes that were made:

### Auth

<details>
<summary>Open to see relevant changes</summary>

- 26ba65f - Initial changes to auth structure
- c7e9892 - Adding the @firebase/auth types

</details>


### Database

<details>
<summary>Open to see relevant changes</summary>

- f16c975 - Initial changes to database structure
- 0c93766 - Adding the @firebase/database types

</details>

### Firestore

<details>
<summary>Open to see relevant changes</summary>

- 5b334fc - Initial changes to firestore structure
- 244ae14 - Adding the @firebase/firestore types

</details>

### Messaging

<details>
<summary>Open to see relevant changes</summary>

- 376d80c - Initial changes to messaging structure
- aa1f7b0 - Adding the @firebase/messaging types

</details>

### Storage

<details>
<summary>Open to see relevant changes</summary>

- fd26887 - Initial changes to storage structure
- 30e8354 - Adding the @firebase/storage types

</details>

There are a couple of cosmetic commits in there and a bugfix or two where an approach wasn't doing what I needed to do. Those commits don't actually change any code in what I have highlighted but move it to locations where everything works as expected.

### Testing Flow

I manually validated these changes doing the following:

1. Run the following in your terminal (at the root of the package):
    `$(npm bin)/lerna exec --scope @firebase/* --scope firebase -- yarn link`
1. Create a separate sandbox directory to test changes
1. Run `yarn link <COMPONENT>` for every component I intend to test

    _e.g._
    ```
    $ yarn link @firebase/app
    $ yarn link @firebase/messaging
    ```
1. Create a stub .ts file and import the components I linked.
1. Exercise the API as I'd expect it to work.

### Testing

Additionally (and I will continue adding to this), I have added a `test` folder to each `@fireabse/*-types` package. In this I can exercise the API as I'd expect and ensure that the file properly compiles. This tests the validity of the typings with regards to our expected use cases. These files will never actually be run so I don't need to make them semantically correct, we simply are using them to test all expected use cases (this is how the typings package [`DefinitelyTyped`](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package) tests the typings for each respective package.
